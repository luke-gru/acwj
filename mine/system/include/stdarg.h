#ifndef _STDARG_H_
#define _STDARG_H_

typedef struct Va_List {
    int gp_offset;
    int fp_offset;
    long overflow_arg_area;
    char *reg_save_area;
} va_list;

#define va_start(list, param) \
    list.gp_offset = 50;\
    list.fp_offset = 0;\
    list.overflow_arg_area = (long)__builtin_vararg_addr_setup();\
    list.reg_save_area = 0

#define va_arg(list, type) ((type)(list.overflow_arg_area = list.overflow_arg_area + 8), (*(type*)(list.overflow_arg_area - 8)))
#define va_end(list)

#endif
