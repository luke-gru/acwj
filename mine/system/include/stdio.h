#ifndef _STDIO_H_
#define _STDIO_H_

#include <stddef.h>
#include <stdarg.h>

#define EOF (-1)

// This FILE definition will do for now
typedef char *FILE;
extern FILE *stderr;
extern FILE *stdout;

FILE *fopen(char *pathname, char *mode);
size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);
size_t fwrite(void *ptr, size_t size, size_t nmemb, FILE *stream);
int fclose(FILE *stream);
int printf(char *format);
int fprintf(FILE *stream, char *format);
size_t getline(char **lineptr, size_t *n, FILE *stream);

FILE *popen(const char *command, const char *type);
int pclose(FILE *stream);

int vfprintf(FILE *stream, const char *format, va_list ap);

#endif	// _STDIO_H_
