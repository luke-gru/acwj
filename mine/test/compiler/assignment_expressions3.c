#include <stdarg.h>
#include <stdio.h>

struct Val {
  int val;
};

struct Val v;

int main() {
  int a = 1;
  a += 1;
  v.val = 0;
  v.val += 1;
  printf("a: %d\n", a);
  printf("v.val: %d\n", v.val);
  return (0);
}

