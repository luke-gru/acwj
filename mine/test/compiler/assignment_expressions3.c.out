a: 2
v.val: 1
